File: Backend/src/main/java/com/meltic/gmao/model/Maquina.java
package com.meltic.gmao.model;

import java.util.ArrayList;
import java.util.List;
import jakarta.persistence.*;

@Entity
@Table(name = "maquinas")
public class Maquina {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String nombre;
    private String modelo; 
    private String descripcion;
    private String ubicacion;
    private String estado; // OK, WARNING, ERROR

    @OneToMany(mappedBy = "maquina", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<MetricConfig> configs = new ArrayList<>();

    public Maquina() {}

    public void addConfig(MetricConfig config) {
        configs.add(config);
        config.setMaquina(this);
    }

    public void removeConfig(MetricConfig config) {
        configs.remove(config);
        config.setMaquina(null);
    }

    // Getters y Setters
    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }

    public String getNombre() { return nombre; }
    public void setNombre(String nombre) { this.nombre = nombre; }

    public String getModelo() { return modelo; }
    public void setModelo(String modelo) { this.modelo = modelo; }

    public String getDescripcion() { return descripcion; }
    public void setDescripcion(String descripcion) { this.descripcion = descripcion; }

    public String getUbicacion() { return ubicacion; }
    public void setUbicacion(String ubicacion) { this.ubicacion = ubicacion; }

    public String getEstado() { return estado; }
    public void setEstado(String estado) { this.estado = estado; }

    public List<MetricConfig> getConfigs() { return configs; }
    public void setConfigs(List<MetricConfig> configs) { this.configs = configs; }
}

File: Frontend/meltic_gmao_app/lib/models/maquina.dart
import 'metric_config.dart';

class Maquina {
  final int? id;
  final String nombre;
  final String ubicacion;
  final String estado;
  final String? modelo;
  final String? descripcion;
  final List<MetricConfig> configs;
  final String? plcUrl;
  final bool simulado;

  Maquina({
    this.id,
    required this.nombre,
    required this.modelo,
    required this.ubicacion,
    this.descripcion,
    required this.estado,
    this.configs = const [],
    this.plcUrl,
    this.simulado = false,
  });

  factory Maquina.fromJson(Map<String, dynamic> json) {
    return Maquina(
      id: json['id'] is int ? json['id'] : null,
      nombre: json['nombre'] ?? '',
      ubicacion: json['ubicacion'] ?? '',
      estado: json['estado'] ?? 'OK',
      modelo: json['modelo'],
      descripcion: json['descripcion'],
      plcUrl: json['plcUrl'],
      simulado: json['simulado'] ?? false,
      configs: (json['configs'] as List?)
              ?.map((c) => MetricConfig.fromJson(c))
              .toList() ?? [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'nombre': nombre,
      'ubicacion': ubicacion,
      'estado': estado,
      'modelo': modelo,
      'descripcion': descripcion,
      'plcUrl': plcUrl,
      'simulado': simulado,
      'configs': configs.map((c) => c.toJson()).toList(),
    };
  }

  Maquina copyWith({
    int? id,
    String? nombre,
    String? ubicacion,
    String? estado,
    String? modelo,
    String? descripcion,
    List<MetricConfig>? configs,
    String? plcUrl,
    bool? simulado,
  }) {
    return Maquina(
      id: id ?? this.id,
      nombre: nombre ?? this.nombre,
      ubicacion: ubicacion ?? this.ubicacion,
      estado: estado ?? this.estado,
      modelo: modelo ?? this.modelo,
      descripcion: descripcion ?? this.descripcion,
      configs: configs ?? this.configs,
      plcUrl: plcUrl ?? this.plcUrl,
      simulado: simulado ?? this.simulado,
    );
  }
}
